### PR TITLE
[SPARK-12506][SPARK-12126][SQL]use CatalystScan for JDBCRelation

### PR DESCRIPTION
**Pull request by huaxingao**
As suggested <a href="https://issues.apache.org/jira/browse/SPARK-9182?focusedCommentId=15031526&amp;page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15031526" class="external-link" rel="nofollow">here</a>, I will change JDBCRelation to implement CatalystScan, and then directly access Catalyst expressions in JDBCRDD.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @liancheng

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 1]
```diff
@@ -455,31 +458,31 @@ class JDBCSuite extends SparkFunSuite
...
-    assert(doCompileFilter(In("col1", Array("jkl"))) === "col1 IN ('jkl')")
-    assert(doCompileFilter(Not(In("col1", Array("mno", "pqr"))))
-      === "(NOT (col1 IN ('mno', 'pqr')))")
-    assert(doCompileFilter(IsNull("col1")) === "col1 IS NULL")
-    assert(doCompileFilter(IsNotNull("col1")) === "col1 IS NOT NULL")
-    assert(doCompileFilter(And(EqualNullSafe("col0", "abc"), EqualTo("col1", "def")))
-      === "((NOT (col0 != 'abc' OR col0 IS NULL OR 'abc' IS NULL) "
-        + "OR (col0 IS NULL AND 'abc' IS NULL))) AND (col1 = 'def')")
-  }
+//  test("compile filters") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: It looks you forgot to uncomment those..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 2]
```diff
@@ -20,14 +20,15 @@ package org.apache.spark.sql.execution.datasources.jdbc
 import java.sql.{Connection, Date, ResultSet, ResultSetMetaData, SQLException, Timestamp}
 import java.util.Properties
 
+import org.apache.spark.sql.catalyst.expressions._
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Fix an import order.

**Comment by maropu** [comment]
Great work though, I think that we should consider lots of things to refactor these kinds of datasource push-down codes because related tickets (SPARK-12449, 12686, ...) describes non-filter logical operators push downs. Also, this pr could break binary compatibility...

**Comment by HyukjinKwon** [comment]
As @maropu said, for me I also agree with him but some may think differently as mentioned [here](https://github.com/apache/spark/pull/10750#issuecomment-175400704).

**Comment by viirya** [comment]
Is `PrunedFilteredScan` and `CatalystScan` conflicting? If not, can we keep original `JDBCRelation` and let it implement `CatalystScan` too? In order not to duplicate current codes, I am wondering if `SQLBuilder` can help converting filtering expressions to SQL string.

**Comment by maropu** [comment]
I agree that `CatalystScan` is used to support arithmetic operations in datasources though, this current `CatalystScan` trait only processes filter expressions. If we make this kind of interface changes, it'd be better to consider other requirements discussed in jira; otherwise, we get stuck with the same issue every time we add new features in datasources.

Anyway, the idea to share codes in `SQLBuilder` and `JDBCRDD` is good to me.

**Comment by rxin** [comment]
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one. We can also continue the discussion on the JIRA ticket.

**Comment by kyprifog** [comment]
@rxin Why was this pull request closed?  Can you direct me to a new one that could have replaced it? I was trying to track down the modification of push downs for basic operations like "limit".

**Comment by HyukjinKwon** [comment]
Technical reason: It's kind of risky to rely on `CatalystScan` and completely replace the interface. I think I already see some tests were disabled here. Also, there look potential better suggestions above.

Practical reason: there are too many pending PRs as you see. If the author is not responsive and the PR is inactive to review comments, we better leave them closed for now - seems it's already stuck in few technical reasons. The author is welcome to reopen and other contributors are welcome to take over.

**Comment by HyukjinKwon** [comment]
BTW, datasource v2 is in progress too to allow more push downs (see [SPARK-22386](https://issues.apache.org/jira/browse/SPARK-22386)). You might want to take a look

**Comment by kyprifog** [comment]
@HyukjinKwon Thanks this is what I was looking for.  Glad the work is still being continued.

**Pull request closed without merging**